__new_qenv.R

```diff
@@ -0,0 +1,316 @@
+    )
+  } else {
+    Filter(function(x) !identical(x, ""), attr(x, condition))
+  }
+}
+
+
+#' @examples
+#'
+#' q <- qenv()
+#' # execute code
+#' with(q, {
+#'   i <- iris
+#'   m <- mtcars
+#' })
+#' q
+#' # error messages are stored
+#' with(q, {
+#'   subset(i, Species == species) # raises error and stops evaluation
+#'   ms <- subset(m, cyl == 4) # not evaluated
+#' })
+#' q
+#' # warnings and messages are also stored
+#' with(q, {
+#'   warning("this is a warning")
+#' })
+#' with(q, {
+#'   message("this is a message")
+#' })
+#' q
+#'
+#' access variables and environment history
+#' q$m
+#' get_code(q)
+#' get_conditions(q, "error")
+#'
+#' # injecting values into code
+#' q <- qenv()
+#' with(q, i <- iris)
+#' with(q, print(dim(subset(i, Species == "virginica"))))
+#' \dontrun{
+#' with(q, print(dim(subset(i, Species == species)))) # fails
+#' }
+#' with(q, print(dim(subset(i, Species == species))), species = "versicolor")
+#' species_external <- "versicolor"
+#' with(q, print(dim(subset(i, Species == species))), species = species_external)
+
+
+# internal; work in progress
+object_info <- function(x) UseMethod("object_info")
+object_info.data.frame <- function(x) sprintf("%d x %d", dim(x)[1], dim(x)[2])              # nolint
+object_info.matrix <- function(x) sprintf("%s, %d x %d", typeof(x), dim(x)[1], dim(x)[2])   # nolint
+object_info.factor <- function(x) sprintf("%d levels, [%d]", length(levels(x)), length(x))  # nolint
+object_info.default <- function(x) sprintf("%s, [%d]", typeof(x), length(x))                # nolint
+
+
+#' @keywords internal
+# internal function to prepare expression(s) for evaluation
+.prepare_code <- function(expr, text) {
+  # This function cannot handle missing arguments, so the caller passes if statements that return NULL if FALSE.
+
+  # Get parent call to use in error messages.
+  parent_call <- match.call(definition = sys.function(2), call = sys.call(2))
+
+  if ((is.null(expr) && is.null(text)) || (!is.null(expr) && !is.null(text))) {
+    stop("specify either \"expr\" or \"text\": ", deparse1(parent_call), call. = FALSE)
+  }
+
+  if (!is.null(expr) && is.character(expr)) {
+    stop("character vector passed to \"expr\": ", deparse1(parent_call), call. = FALSE)
+  }
+
+  # Add braces to expressions. Necessary for proper storage of some expressions (e.g. rm(x)).
+  if (!is.null(expr) && !grepl("^\\{", deparse1(expr))) {
+    expr <- call("{", expr)
+  }
+
+  # Process compound expression string: split lines, remove braces and white space, ignore empty strings.
+  if (!is.null(text) && length(text) == 1L && grepl("^\\{", text)) {
+    text <- strsplit(text, split = "\n")[[1]]
+    text <- trimws(text, whitespace = "[ \t\r\n\\{\\}]")
+    text <- Filter(Negate(function(x) identical(x, "")), text)
+  }
+
+  code <-
+    if (is.null(text)) {
+      # Drop strings from compound expressions.
+      Filter(Negate(is.character), as.list(expr)[-1])
+    } else if (is.null(expr)) {
+      text
+    }
```

Comment:
A proposal for integrating the `str2expression`

```suggestion
  expr_text <- NULL

  if (!is.null(text)) {
    expr_text <- str2expression(text)

    if (length(expr_text) == 1 && isTRUE(expr_text[[1]][[1]] == "{")) {
      expr_text <- expr_text[[1]][seq(2, length(expr_text[[1]]))]
    }
  }

  code <-
    if (is.null(text)) {
      # Drop strings from compound expressions.
      Filter(Negate(is.character), as.list(expr)[-1])
    } else if (is.null(expr)) {
      Filter(Negate(is.character), as.list(expr_text))
    }
```